rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+use with the following syntax:
+
+```python
+from tvm import meta_schedule as ms
+
+design_space_generator = ms.PostOrderApply(rules=[
+    ms.MultiLevelTiling(...),
+    CustomRule(...),
+    ms.OtherBuiltinRules(...),
+])
+
+```
+
+### 3.5. Unifying manual schedule / AutoTVM / Ansor
+
+In this section, we show that the design space induced by TE manual schedule, AutoTVM and Ansor are
+all subsets of meta schedule, and meta schedule further allows mixing those three styles to search
+jointly.
+
+**Manual schedule**. The TE schedule is a special case of a meta schedule program, where there is no
+randomness introduced by sampling instructions. It is a single point in terms of design space.
+
+**AutoTVM (Template-based tuning)**. It is more natural representation of AutoTVM’s schedule
+templates (knobs) by writing one or more schedule functions in meta schedule with sampling
+instructions. The execution traces generated by the schedule functions are the design space to be
+explored.
+
+**AutoScheduler (Ansor, Template-free tuning)**. As mentioned in the previous section, application
+of composite schedule rules generates the design space, which is equivalent to Ansor’s sketch
+generation.
+
+**Mixing styles in design space definition**. By taking union of the spaces induced by the three
+special cases, our system allows developers to combine generic rules that Ansor provides and
+operator-specific scheduling.
+
+## 4. Reference-level explanation
+
+In this section, we introduce the underlying techniques for the automation system to extract and
+explore the design space. The figure below briefly illustrates the workflow of the system:
+
+![meta-schedule-workflow](../resources/meta-schedule-workflow.png)
+
+### 4.1. Execution trace as the design space
+
+**Trace**. To represent the design space defined by the meta schedule DSL, the underlying system
+records all the instructions users applied to the schedule class, including sampling and schedule
+primitives. We call this list of instructions a trace.
+
+For instance, executing the example above results in the following trace:
+
+```
+Instruction 0. Sample-Perfect-Tile
+Instruction 1. Sample-Perfect-Tile
+Instruction 2. Sample-Perfect-Tile
+Instruction 3. Split
+Instruction 4. Split
+Instruction 5. Split
+Instruction 6. Reorder
+```
+
+**Trace forms design space.** A trace may contain zero or more sampling instructions, which
+introduces the uncertainty in scheduling - one instance of sampling results in one point in the
+design space. Therefore, the trace itself forms a design space to explore, e.g. which set of tile
+sizes works best on a specific hardware.
+
+**Union of design space**. Our system works on a set of traces, representing the union of the design
+spaces represented by every single trace.
+
+**Fork a trace**. When two different decisions in the scheduling process are equally important to
```

Comment:
how would we know they are "equally important?" do you mean that one decision may yield better performance on one architecture, and another decision may yield better performance on another architecture?

Reply:
For example, in a `conv2d-relu` subgraph, when fusing relu into the loop nests of conv2d, we need to decide the fusion level, i.e. should i compute relu on the second tile or the third one? In this case, these decisions are equally important